### PR TITLE
ACC-1347 Create metrics for syndication article and video downloads

### PR DESCRIPTION
**Pull request by asugar13**
### Description
Creating graphite metrics for Syndication article and video downloads
### Ticket
[JIRA](https://financialtimes.atlassian.net/jira/software/c/projects/ACC/boards/1108?modal=detail&selectedIssue=ACC-1347)

### CODE REVIEW

File: server/controllers/download-by-content-id.js [hunk 1]
```diff
@@ -99,6 +101,24 @@ module.exports = exports = async (req, res, next) => {
...
+		log.count('article-download-start')
+
+		dl.on('error', (err) => {
+			log.error('DOWNLOAD_ARTICLE_ERROR', {
+				event: 'DOWNLOAD_ARTICLE_ERROR',
+				error: err.stack || err
+			});
+			log.count('article-download-error');
+			res.status(500).end();
+		});
```
Comment (joelcarr) [inline server/controllers/download-by-content-id.js:113]: **Suggestion:** As this is the same as the `dl.on('error'...)` event in the `if` statement, we can move this above the if/else statement only needing to write it once.

```javascript
	log.count('archive-download-start');
	
	dl.on('error', (err) => {
		log.error('DOWNLOAD_ARTICLE_ERROR', {
			event: 'DOWNLOAD_ARTICLE_ERROR',
			error: err.stack || err
		});
		log.count('article-download-error');
		res.status(500).end();
	});
	
	if (dl.downloadAsArchive) {
		...
	} else {
		...
	}
```

**Review by joelcarr — COMMENTED** [review]
Looks good! I have left a couple suggestions to remove duplicated code but these are non-blocking. Let me know what you think 👍

File: server/controllers/download-by-content-id.js [hunk 2]
```diff
@@ -65,13 +65,15 @@ module.exports = exports = async (req, res, next) => {
 	req.connection.on('close', () => dl.cancel());
 
 	prepareDownloadResponse(res, content);
-
 	if (dl.downloadAsArchive) {
+		log.count('archive-download-start');
```
[the remark this answers is not in the document]
Comment (asugar13) [reply server/controllers/download-by-content-id.js:69]: Good suggestion! However I'm triggering different events depending on whether the download that started is `article` or `archive` a.k.a video format. They are called 'archive-download-start' and 'article-download-start'

File: server/controllers/download-by-content-id.js [hunk 3]
```diff
@@ -81,12 +83,12 @@ module.exports = exports = async (req, res, next) => {
 
 			if (dl.cancelled !== true) {
 				res.end();
-
 				next();
 			}
 		});
 
 		dl.on('complete', (state, status) => {
+			log.count('archive-download-complete')
```
[the remark this answers is not in the document]
Comment (asugar13) [reply server/controllers/download-by-content-id.js:91]: Great spot. Updating.

File: server/controllers/download-by-content-id.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (joelcarr) [reply server/controllers/download-by-content-id.js:69]: Ahh yes! Sorry my poor reading there. That makes complete sense!

File: server/controllers/download-by-content-id.js (hunk 1 above)
Comment (joelcarr) [reply server/controllers/download-by-content-id.js:113]: There is a difference in the two callbacks, this suggestion would not work for our need. Resolving this conversation.

File: server/controllers/download-by-content-id.js [hunk 4]
```diff
@@ -66,34 +66,40 @@ module.exports = exports = async (req, res, next) => {
...
-				event: 'DOWNLOAD_ARCHIVE_ERROR',
-				error: err.stack || err
-			});
-
-			res.status(httpStatus || 500).end();
-		});
-
-		dl.on('end', () => {
-			log.debug(`DownloadArchiveEnd => ${content.id} in ${Date.now() - dl.START}ms`);
+	let articleOrArchive = dl.downloadAsArchive ? 'Archive' : 'Article'
```
Comment (joelcarr) [inline server/controllers/download-by-content-id.js:69]: **Praise:** NICE solution! 🙌

Comment (joelcarr) [reply server/controllers/download-by-content-id.js:69]: **Nitpick:** Please could we add a `;` to the end of the line?

File: server/controllers/download-by-content-id.js [hunk 5]
```diff
@@ -66,34 +66,40 @@ module.exports = exports = async (req, res, next) => {
...
 
-			if (dl.cancelled !== true) {
-				res.end();
+	log.count(`${articleOrArchive.toLowerCase()}-download-start`)
```
Comment (joelcarr) [inline server/controllers/download-by-content-id.js:71]: **Nitpick:** Please could we add a `;` to the end of the line?

File: server/controllers/download-by-content-id.js [hunk 6]
```diff
@@ -66,34 +66,40 @@ module.exports = exports = async (req, res, next) => {
...
+		log.count(`${articleOrArchive.toLowerCase()}-download-error`);
+		res.status(500).end();
+	});
 
-		dl.on('complete', (state, status) => {
-			res.status(status);
-		});
+	dl.on('complete', (state, status) => {
+		if (state === 'complete') {
+			log.count(`${articleOrArchive.toLowerCase()}-download-complete`)
```
Comment (joelcarr) [inline server/controllers/download-by-content-id.js:84]: **Nitpick:** Please could we add a `;` to the end of the line?

**Review by joelcarr — COMMENTED** [review]
Fantastic refactor Asier! Just a few very nitpicky comments about adding semi colons to the end of lines.

File: server/controllers/download-by-content-id.js (hunk 5 above)
Comment (asugar13) [reply server/controllers/download-by-content-id.js]: done

File: server/controllers/download-by-content-id.js (hunk 4 above)
Comment (asugar13) [reply server/controllers/download-by-content-id.js]: done

File: server/controllers/download-by-content-id.js [hunk 7]
```diff
@@ -81,25 +81,25 @@ module.exports = exports = async (req, res, next) => {
...
-			next();
-		}
-	});
-
 	dl.on('cancelled', () => {
 		next();
 	});
 
 	if (articleOrArchive === 'Archive') {
+		dl.on('end', () => {
```
Comment (asugar13) [inline server/controllers/download-by-content-id.js:94]: I've tested that code (`dl.on`) and it only runs for archives (a.k.a videos).  So I'm moving to the `if` clause

**Review by joelcarr — APPROVED** [review]
Good work Asier!

**Pull request merged**